dev/io.openliberty.data.internal.persistence/src/io/openliberty/data/internal/persistence/QueryInfo.java

```diff
@@ -221,11 +221,22 @@ public static enum Type {
     final Class<?> returnArrayType;
 
     /**
-     * The type of a single result obtained by the query.
-     * For example, a single result of a query that returns List<MyEntity> is of the type MyEntity.
+     * The type of a single result obtained by the query. For example,
+     * A query that returns List<MyEntity> has singleType MyEntity.
+     * A query that returns List<ArrayList<String>> has singleType ArrayList<String>.
+     * A query that returns Optional<String[]> has singleType String[].
      */
     final Class<?> singleType;
 
+    /**
+     * Element type of singleType when singleType is an array or collection.
+     * Null if singleType is not an array or collection. For example,
+     * A query that returns List<MyEntity> has singleTypeElementType null.
+     * A query that returns List<ArrayList<String>> has singleTypeElementType String.
+     * A query that returns Optional<String[]> has singleTypeElementType String.
+     */
+    final Class<?> singleTypeElementType;
```

Comment:
suggestion (non-blocking): The name of this field is confusing to me, but the javadoc clears it up.  Perhaps a better name would be `singleTypeCollectionType`?

Reply:
It's difficult to fine a decent name.  In the last example given, `Optional<String[]>` with singleTypeElementType of `String`,`String[]` is an array, not a Collection.